### PR TITLE
Fix#8 Add more distributions

### PR DESCRIPTION
**Pull request by ZEXINLIU**
Add implements for Laguerre and Hermite polynomials.
Add induced distribution for half-freud and freud weights.

To be done:
Add inverse for each of these distributions.
Add fast algorithms for each of them.

### CODE REVIEW

**Comment by akilnarayan** [comment]
Also an implementation of `ExponentialDistribution` and `NormalDistribution` in `distributions.py` should be done. Basic multivariate functionality for now. We can do more complicated multivariate stuff later.

**Comment by akilnarayan** [comment]
Looks good, I fixed some issues, including the close-to-0 zero for Hermite polynomials. 

Still needs the `ExponentialDistribution` and `NormalDistribution` classes, then I think it'll be ready for merging

File: distributions.py [hunk 1]
```diff
@@ -9,6 +9,331 @@ class ProbabilityDistribution:
...
+            elif len(mean) == 1 and len(stdev) > 1:
+                mean = [mean[0] for i in range(len(stdev))]
+            
+            elif len(stdev) == 1 and len(mean) > 1:
+                stdev = [stdev[0] for i in range(len(mean))]
+
+        elif meaniter: # mean is iterable, stdev is not
+            if stdev is None:
+                stdev = 1.
+            stdev = [stdev for i in range(len(alpha))]
```
Comment (akilnarayan) [inline distributions.py:99]: What is alpha here?

File: distributions.py [hunk 2]
```diff
@@ -9,6 +9,331 @@ class ProbabilityDistribution:
...
+        Sets self.dim, self.mean, self.stdev, self.cov
+        """
+        # Situations:
+        # 1. User specifies mean, stdev lists and cov array (disallow contradictory
+        #    dimension specification)
+        # 2. User specifies dim scalar (disallow contradictory mean, stdev and
+        #    cov specification)
+        # 3. dim = 1
+
+        if len(alpha) > 1 or len(beta) > 1 or cov.shape[0] > 1: # Case 1:
```
Comment (akilnarayan) [inline distributions.py:130]: alpha? beta?

File: distributions.py [hunk 3]
```diff
@@ -9,6 +9,331 @@ class ProbabilityDistribution:
     def __init__(self):
         pass
 
+class NormalDistribution(ProbabilityDistribution):
+    def __init__(self, mean=None, stdev=None, cov=None, dim=None):
+
+        mean,stdev = _convert_meanstdev_to_iterable(mean, stdev)
+        cov = _convert_cov_to_iterable(cov)
+        
+        self._detect_dimension(mean, stdev, cov, dim)
```
Comment (akilnarayan) [inline distributions.py:18]: I think we can just do `cov` as an input. The input `stdev` is not necessary, and would probably be confusing.

In addition, a standard deviation (`stdev`) is not really well-defined for a multivariate non-isotropic Gaussian. (Is it the Cholesky factor of `cov`? Is it the spd square root of `cov`? Is it the square root of the diagonal of `cov`?

File: families.py [hunk 4]
```diff
@@ -29,6 +29,7 @@ def jacobi_recurrence_values(N, alpha, beta):
                       sp.gammaln(alpha + 1.) + sp.gammaln(beta + 1.) -
                       sp.gammaln(alpha + beta + 2.)
                     )
+        ab = np.sqrt(ab)
```
Comment (akilnarayan) [inline families.py:32]: Why this change? Was the `sqrt` not there before? This change doesn't seem relevant to the exponential/normal distributions.

File: distributions.py [hunk 5]
```diff
@@ -9,6 +9,331 @@ class ProbabilityDistribution:
...
+        
+        elif lbd is None:
+            lbd = [1,]
+        
+        else: # lbd is a scalar
+            lbd = [lbd,]
+
+        return lbd
+
+    def meanstdev_to_lbd(self, mu, stdev):
```
Comment (akilnarayan) [inline distributions.py:308]: This looks to just make sure the parameters match an exponential distribution on [0, \infty). But we should allow users to build in affine maps. E.g., for any mu, stdev (>0), you can compute the affine map that would take that exponential distribution to the lbd=1 distribution on [0, \infty). 

There should also be some sort of flag to indicate whether the distribution is reflected or not. I.e., is it a distribution on (-\infty, 0] or [0, \infty)? Specifying `mu` and `stdev` is not enough to infer this, so we'll need an extra input.

Anyway, all of this would just be used to define the `AffineTransform`s in the constructor, and not much else.

**Comment by dcwhite** [comment]
@akilnarayan And of course this one will need a big painful merge after merging your file move PR. It's already conflicted; should it be redone after the sphinx/file move one is merged in?

**Comment by akilnarayan** [comment]
I'd already chatted with Zexin about this -- I think the strategy we were going to use is to merge master into this branch once the docs/sphinx are in master. That should allow us to do the painful merging locally.

This merge in principle shouldn't be that bad since the affected files are mostly disjoint from the docs, but I just know that things will be worse than expected. 

So that was the plan -- do you think there is an easier way?

**Comment by dcwhite** [comment]
That sounds good. I’ll approve the Sphinx one today.

**Comment by dcwhite** [comment]
The build is failing due to imports. Running a full set of flake8 tests, it sees the import but doesn't think it actually loads the name:
```
UncertainSCI/opolynd.py:2:1: F401 'UncertainSCI.opoly1d' imported but unused
...
UncertainSCI/opolynd.py:25:17: F821 undefined name 'opoly1d'
```

**Review by akilnarayan — APPROVED** [review]
